### DISCUSSION TITLE
Artisan v3.1.1 (704a1f8, 0) An error occurred after clicking CHARGE

### DISCUSSION THREAD

**Discussion by Terracotta-6** [artisan-roaster-scope/artisan#1764] [General]

**Comment by MAKOMO** [comment]
Yep, this is a known issue of the underlying pymodbus lib which we just update the trunk to. That will be fixed upstream and we will upgrade pymodbus again once available. Please use the release v3.1 for now until this is fixed.

As always: THANKS FOR YOUR REPORT!

**Reply by MAKOMO** [reply]
See https://github.com/pymodbus-dev/pymodbus/pull/2503

**Reply by MAKOMO** [reply]
Should be fixed by my latest commit (upgrade to pymodbus 3.8.1). New continuous builds should be online in about 30min.

**Reply by Terracotta-6** [reply]
This issue has been fixed thanks

**Comment by Terracotta-6** [comment]
File "artisanlib\main.py", line 1386, in run File "artisanlib\main.py", line 10371, in eventaction_internal File "logging\_init__.py", line 1574, in exception File "lo...

**Reply by MAKOMO** [reply]
Could you quickly try out the current build. This should resolve this issue (introduced by GitHubs precommit itself:(

**Reply by Terracotta-6** [reply]
This issue has been fixed thanks

**Comment by Terracotta-6** [comment]
Artisan v3.1.1 (2380f05, 0) eventsliders write register: Modbus Command problem, Control gets messy results.

**Reply by MAKOMO** [reply]
Need more details. What is the definition of the slider? What is the result?

**Reply by Terracotta-6** [reply]
Previous version Artisan v3.1.1 (2380f05, 0) didn't have this problem when using sliders.
The temperature controller panel shows confusing results, artisan does not report an error

**Reply by MAKOMO** [reply]
Use the more specific `writeSingle` (writes always one register integer) instead of `write` (which decides based on its argument type to write a one register integer or a 2 register float). A change I applied does now feed slider values as floats. The next commit will fix this and again send integers in your case. However, the new ramping event slider actions will send floats. In this case you need to use `writeSingle` which rounds its argument to the nearest integer value send to one register or explicitly use `writeWord` to write floats to 2 registers. THANKS FOR THE NOTICE!

**Reply by Terracotta-6** [reply]
Writing the register in the writeSingle format is valid and the problem is solved. Thanks for your answer!

**Reply by Terracotta-6** [reply]
This has been fixed in Artisan v3.1.1 (a722fae, 0)

**Comment by Terracotta-6** [comment]
Drag the slider with the mouse is not smooth can be what reason?

**Reply by MAKOMO** [reply]
Is this a new problem with the build https://github.com/artisan-roaster-scope/artisan/commit/a722fae3d31b50b2965926d629c606010ab1bec7 of today? In any case, I need your artisan-settings.aset to see how things are configured.

**Reply by Terracotta-6** [reply]
No, it was there in the past, because I don't use the slider to adjust the SV. artisan-settings.aset sent via email.

**Comment by Terracotta-6** [comment]
"button((),false)" Invalid for Modbus Command.
![6a77e437967aff84aa13ef94574fbce](https://github.com/user-attachments/assets/db5665d6-eca0-4fbe-af45-fb35b1f348dd)

**Reply by MAKOMO** [reply]
button() is a Artisan Command not a Modbus Command. For documentation of commands see the help button in the same dialog.

**Reply by MAKOMO** [reply]
Upps. Yes, indeed. `button` is also a valid MODBUS command, but it does not work. Will fix this.

**Reply by MAKOMO** [reply]
Thanks for reporting. Fix is commited and new build is being produced by CI and should be online within the next 30min.

**Reply by Terracotta-6** [reply]
Thank you very much for your efficient work! Button status display works fine.

**Comment by Terracotta-6** [comment]
Extended background curve Some parameter adjustments are not displayed with parameter changes and are displayed normally in the foreground.
![078275e87c92fd84255bcf277856d66](https://github.com/user-attachments/assets/6dd7b445-af5a-41eb-918a-a2baa02f0ea9)

**Reply by MAKOMO** [reply]
I might not understand your issue correctly, but the background curves are here listed with prefix "Background" like "BackgroundET", "BackgroundBT", "BackgroundXT" and "BackgroundYT" (not only two extra curves can be display in the background as configured in the Background dialog.

**Reply by Terracotta-6** [reply]
https://github.com/user-attachments/assets/29ff15e2-130e-4224-92d7-923e1b2c1ade

The extended curve line width in the background curve does not show the actual line width.

**Comment by Terracotta-6** [comment]
build -4fdb284   
Detected :Trojan:Win32/Wacatac.B!ml

**Reply by roasterdave** [reply]
This is a false positive.  Microsoft Defender gets trigger happy from time to time.  Nothing we can do on the Artisan side.  This build differs from the previous build by just one help file.   The  other two files in the commit are source files that are not packaged in the installer.

You can add artisan.exe to the Defender Exclusion list to prevent this action.

**Reply by Terracotta-6** [reply]
Thank you for your reply! Does not affect my use.

**Comment by Terracotta-6** [comment]
Background events cannot be updated to ladder graph "ramp" when using "Artisan Command ramp(n,<bool>)".

**Reply by MAKOMO** [reply]
A redraw was missing. Will be fixed with my next commit. THANKS for reporting!

**Reply by Terracotta-6** [reply]
Artisan Command ramp is ready. Thank you!

**Reply by Terracotta-6** [reply]
What caused the problem?
Exception: playbackevent() list index out of range@line 5884
This has been fixed .Thank you!

**Reply by MAKOMO** [reply]
Very good find. Cause is that Artisan tries here to replay one of the "unnamed" event of type "--". That does not work. I have a fix coming and will trigger a new build in a minute.

**Comment by Terracotta-6** [comment]
In an ramp, lose the last clip, I don't understand the ramp control rules， Did I violate the ramp rules?
![6add11840e703e37358ae4fdff2930c](https://github.com/user-attachments/assets/9a12408c-c0ac-4225-85a2-cb79074394c6)

**Reply by MAKOMO** [reply]
Could you please provide your settings and background file with instructions on how to replicate what you are doing. Then I might be able to answer your question.

**Comment by Terracotta-6** [comment]


**Reply by MAKOMO** [reply]
The ramping mechanism is calling the slider actions on intermediate values interpolated between two events. It is hard to see from your original photo, but it seems that ramping first works after an event has be replayed and mid-way it stops. Very hard to understand how you generate those intermediate (ramping) events as by default Artisan does not generate them, but just issues the slider action. Could you  share your complete artisan configuration as settings file to understand what I see here?

**Reply by Terracotta-6** [reply]
Only the last paragraph is not played in the ramp mode.
![dce651ccd5060c7e3b3d212f0875422](https://github.com/user-attachments/assets/04fdf305-37e8-480a-8bfb-45ea1bdbff13)

**Reply by MAKOMO** [reply]
I understand. Artisan only replays events between CHARGE and DROP. Your last event seems to be exactly at DROP or short after and thus is not replayed. However, it should be used to define this last ramp. I will try to improve on this. Thanks for reporting!

**Comment by MAKOMO** [comment]
But if I understand correctly, those other initial parts of the two marked ramp are also not re-played, right?

![431807910-9a12408c-c0ac-4225-85a2-cb79074394c6-1](https://github.com/user-attachments/assets/55a2433c-56e1-4d59-bdd0-a7b561618ba4)

**Reply by Terracotta-6** [reply]
No problem showing ramp playback on the thermostat panel .

**Comment by MAKOMO** [comment]
Thanks for your report! I did fix this and replay should now also ramp to the next event if this is on DROP or later. To be commited....

**Reply by MAKOMO** [reply]
I will revert this change as it can lead to unexpected behavior (as it just happend to me). After DROP you often have an event that turns down the heater completely to cool down. If the last event before DROP ramps down to such an event the heater level will drop down quickly BEFORE DROP which is for sure not what is intended. Thus events to be ramped towards need to be set before (or exactly at) DROP time. Sorry.

**Reply by Terracotta-6** [reply]
I tested the build c24a557, playing the last segment of the ramp in its entirety, and found no other issues. Respect your choice. Thank you!
